homeassistant/components/generic/camera.py

```diff
@@ -120,18 +120,15 @@ def frame_interval(self):
         """Return the interval between frames of the mjpeg stream."""
         return self._frame_interval
 
-    def camera_image(
-        self, width: int | None = None, height: int | None = None
-    ) -> bytes | None:
-        """Return bytes of camera image."""
-        return asyncio.run_coroutine_threadsafe(
-            self.async_camera_image(), self.hass.loop
-        ).result()
-
     async def async_camera_image(
         self, width: int | None = None, height: int | None = None
     ) -> bytes | None:
         """Return a still image response from the camera."""
+        if not self._still_image_url:
+            if self.stream:
+                return await self.stream.async_get_image(width, height)
+            await self.async_create_stream()
```

Comment:
What does this call do? It seems like it will create the stream, but not start it, is that right? It seems like the next call to async_get_image will be safe, but will just return None, unless something else starts the stream, if i understand right.

Reply:
Yes, you are right, we should start the stream there too

Reply:
This is a reasonably different behavior change, so i think this could also come in a follow up PR to evaluate on its own, but for now using the image if already preloaded since like a nice win.

Reply:
I already pushed the relevant change in b5b26bf

Reply:
Rebased to try and fix the CI problem.
I'm curious as to what you think of the `stream.start()` in here. It might make more sense to move that into `Stream.async_get_image`

Reply:
Just moved that as well. I know this PR is touching a few things at once, but as it's the first usage of the camera <-> Stream.async_get_image interface we should try to fix things up to get it right.
Let me know if you want me to break up the PR.

Reply:
I have few cases I am thinking about here:
(1) If the stream fails and is not restarted, this will show the last image. The current ffmpeg call will get a fresh frame every time.
(2) Creating the stream without starting it will just get back `None`, so the fallback to ffmpeg would not longer happen
(3) Not yet sure how I feel about always starting the stream, but it would fix those issues at the cost of having the stream always on. That seems like a larger topic that may need more considerations, data, etc. (e.g. maybe we collect data about the impact of such a change to argue whether or not its ok? or maybe showing that the ffmpeg frame thing is super expensive, etc)

Another alternative that is more incremental could be to only use the stream if its created /and/ active (or only use the stream image with X minutes before it is considered stale if a stream failed) before falling back to ffmpeg. This is extra code, but preserves status quo behavior.

Curious to hear your thoughts on this. I definitely like the direction of trying to replace ffmpeg.  This also makes me think it would be really cool to be able to collect stats w.r.t. stream bandwidth usage.

Reply:
I guess I have two ways of looking at this.
1) From the generic component point of view:
Currently the generic component uses `CONF_STILL_IMAGE_URL` to get still images and `CONF_STREAM_SOURCE` to get the video. For all intents and purposes, these two are just distinct concepts that are tied in the same entity. It's unclear for many users why they have to put in two different URLs to access the same device (and even the username and password concepts are very unclear - see #23064 and #45510). What's worse, some cameras do not even have a `CONF_STILL_IMAGE_URL`. This PR, including starting the stream on demand, would allow `generic` to work seamlessly with just a `CONF_STREAM_SOURCE` and picture entity cards. Without starting the stream on demand, the images would just come back None.
2) From camera components using haffmpeg:
This would replace reconnecting to a camera every 10 seconds with being constantly connected to the camera. It's debatable which is preferred as they have pros and cons:
ffmpeg pros: less bandwidth between the camera and HA overall, since it only grabs keyframes at intervals and nothing in between.
stream pros: more bandwidth between the camera and HA overall since the rtsp stream is still connected between keyframes. However, there is less load on the camera as the image generation is moved from the camera to HA. There's also less overhead for a new ffmpeg process and RTSP setup every time an image is requested (this is not just the protocol setup, but the encoder must start over for each new RTSP connection since it usually starts with a keyframe). Many cameras are limited to only a few simultaneous connections and require some time to reset their RTSP connections, so there may already be some problems when using the existing method. This problem is compounded when several clients are requesting images at the same time, since there's a new ffmpeg process/RTSP connection opened for each image requested - the stream way will scale much better.
I think the automatic stream starting behavior makes sense as it's not really that much more intensive than making a request, unless HA to camera bandwidth is expensive. This is not the case for most cameras, only cloud ones. I guess cloud ones can stick to using the existing ha-ffmpeg method.
 
BTW, if we stick with the automatic starting we may need may need to add a new provider to make use of the IdleTimer / idle_callback.

Reply:
Your point about local cameras makes a lot of sense. I think it is fair to assume that all cameras using this integration are using local RTSP urls.

The only other concern could be CPU usage on a raspberry pi or something.  Do you have any numbers to put behind this? I would assume the overhead of starting pyav is going to be smaller than forking ffmpeg, and it will be more consistent, but maybe having some data to back this up could help.

Reply:
How about we give the stream a 15 second timeout when we start it this way? Since the normal cards grab images ~10 seconds apart, if we don't get 2 consecutive requests then we can stop the stream. Or maybe that's too quick, as the user might switch between views but still want to come back. Maybe a 2-3 minute timeout makes sense.
The current IdleTimer/Provider framework is based on StreamOutput. I'll have to see if we can expand it to work with KeyFrameConverter. Alternatively, maybe we actually want to start up a HLS provider, since a natural use case is to use an "auto" picture entity card, and the regular picture load will help start the HLS provider before the user clicks in to the individual camera.
I don't have any numbers, but my own instance has the rough CPU power of a RPi, and I haven't seen any issues with 10+ cameras refreshing every ~30 secs or so (I use very long keyframe intervals so 2 out of 3 requests get back the old image, so only one is actually doing work).
The CPU usage of the open stream is very low (as we know from preload stream - and if we don't use a HLS provider it's even lower), so the concern would be decoding and encoding. Although we're decoding and encoding using different codecs, encoding is usually a lot more resource intensive than decoding. On this front libjpeg-turbo seems to be very fast - better than PIL which is used in some other integrations. Also, I'm not sure how jpeg scaling works, but the naive assumption is that if any cameras are rescaling to use the new width and height parameters, they'd have to do a decode/reencode as well, so the CPU load would actually be similar to this.

Reply:
- I like your 2-3 minute timeout since we expect the frontend to be polling fairly often.
- I see what you mean about IdleTimer + StreamOutput. I can see that how Stream interacts w/ StreamOutput is very similar for this case (e.g. checking for idle timeout, shutting down, etc)..
- But the side where the worker pushes data to it is different.   I could see how KeyframeConverter could be a StreamOutput, then the StreamMuxer would beed to know about different types of stream outputs (some that want just keyframes pushed in, or some that want Segments) but they could also just ignore the parts they don't care about, or an optimization could be to only do segmentation when the outputs are actually listening for segments, which could come later.

I am not sure if generalizing the idle time out side or stream output side is easier, but may be worth thinking through both to decide.

On performance, thanks for talking it through for me 👍🏼

Reply:
Can this go in as is?

Reply:
Did you look at the most recent push? I ended up just starting up the HLS output as that at least will start the IdleTimer to time everything out. I think it can go in as is now, and we can continue the IdleTimer/StreamOutput discussion.

Reply:
I see, i missed that part. Yes, that looks like it covers it fine, handling the error conditions.